integration/network/service_test.go

```diff
@@ -321,6 +321,70 @@ func noServices(client client.ServiceAPIClient) func(log poll.LogT) poll.Result
 	}
 }
 
+func TestServiceWithDataPathPortInit(t *testing.T) {
+	skip.If(t, testEnv.OSType == "windows")
+	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"), "DataPathPort was added in API v1.40")
+	defer setupTest(t)()
+	var ops = []func(*daemon.Daemon){}
+	var datapathPort uint32 = 7777
+	ops = append(ops, daemon.WithSwarmDataPathPort(datapathPort))
+	d := swarm.NewSwarm(t, testEnv, ops...)
+
+	cli := d.NewClientT(t)
+	defer cli.Close()
+
+	// Create a overlay network
+	name := "saanvisthira" + t.Name()
+	network.CreateNoError(t, context.Background(), cli, name,
+		network.WithDriver("overlay"))
+
+	var instances uint64 = 1
+	serviceName := "TestService" + t.Name()
+	serviceID := swarm.CreateService(t, d,
```

Comment:
I'm actually trying to understand this test; where do we test if the data-port is actually set? Why do we start services here (it looks to be only testing that we can create  a service, but I don't see how this would show that the new feature works).

Should we (instead?) check if the `/swarm` endpoint returns the correct port-number?


We'd also need a test to check if a node that joins the swarm correctly copies the port number (but that would probably have to be a test in the e2e)

Reply:
when we create overlay (VXLAN)  network , we create VXLAN interface which uses dest port as one of input param. But in our swarm, vxlan network gets created only when we create a service using  overlay network ( lazy creation). So I tried to create service and make sure it succeeds. If the port is in use already , then vxlan interface creation will fail and eventually service create also will fail.

Reply:
also checking swarm endpoint would return the value what we stored in RAFT cluster object. we need to make sure we create service with port change to make sure it works at libnetwork level.

Reply:
Right, but currently it's more a "regression test"; we're testing that with this feature we're still able to create services, but we don't test;

- if the port we specified is actually stored in RAFT (which we should test by checking the `/swarm` endpoint)
- related to the above; if the feature doesn't work correctly, and (e,g,) we initialize with port 1234, but swarm ignores this, and uses the default port (4789), the test still passes 
- if the service / network is actually using the port we specified (and what's stored in RAFT)


For the last one, not sure how to test; a really dirty way would be to;

- init swarm with a port (eg 5678)
- create a container that publishes port 5678
- create a service

the service should fail to be deployed, because port 5678 is already in use by the container (but we would likely have to check the error to see if that's the reason it failed, and not any other reason)

Reply:
We have checks in every module  which  will return error if the port configuration fails during swarm init.  if swarn init with  datapath port succeeds, then we have successfully configured all the way in libnetwork.
To answer your point "we initialize 1234 but swarm takes 4789" - I think we can write unit test in swarmkit RAFT read/write and calling libnetwork code path to make sure swarmkit is working properly. 
For the last one : That doesn't look like the correct way of testing. 

But I will see if I can expand the current test.

Reply:
> We have checks in every module which will return error if the port configuration fails during swarm init. if swarn init with datapath port succeeds, then we have successfully configured all the way in libnetwork.

We're currently only testing the "happy path", so we didn't verify if those checks do their job (e.g., there's no test trying to initialise with port `1023` or `49152` - which are "out of range", nor do we have a test for a situation where the specified port overlaps with a port that's already in use) 

> To answer your point "we initialize 1234 but swarm takes 4789" - I think we can write unit test in swarmkit RAFT read/write and calling libnetwork code path to make sure swarmkit is working properly.

unit tests in swarmkit and libnetwork are always good to have (if it's calling libnetwork code, it wouldn't be a unit test though 🤔)

_but_ that would not cover the changes in this PR; given that we process the value in this PR (`getDataPathPort()` can adjust the port if it sees it's `0`); we still need a test here that checks we get the correct port.

> For the last one : That doesn't look like the correct way of testing.

Agreed; so we may need another way to verify it's using the correct port